File: lib/widgets/audible_stars.dart
import 'package:flutter/material.dart';

import '../core/audible_rating_service.dart';

/// Shows a cached Audible star rating for a book. Displays the cached value
/// immediately and refreshes in place when it's older than 24h
/// (stale-while-revalidate). Renders nothing until a confident rating exists.
class AudibleStars extends StatefulWidget {
  const AudibleStars({
    super.key,
    required this.itemId,
    required this.title,
    this.asin,
    this.author,
    this.narrator,
    this.durationMs,
    this.region = 'us',
    this.starSize = 16,
    this.color,
    this.showCount = true,
    this.alignment = MainAxisAlignment.start,
  });

  final String itemId;
  final String title;
  final String? asin;
  final String? author;
  final String? narrator;
  final int? durationMs;
  final String region;
  final double starSize;
  final Color? color;
  final bool showCount;
  final MainAxisAlignment alignment;

  @override
  State<AudibleStars> createState() => _AudibleStarsState();
}

class _AudibleStarsState extends State<AudibleStars> {
  AudibleRating? _rating;

  @override
  void initState() {
    super.initState();
    _rating = AudibleRatingService.instance.peek(widget.itemId);
    _load();
  }

  @override
  void didUpdateWidget(AudibleStars old) {
    super.didUpdateWidget(old);
    if (old.itemId != widget.itemId) {
      _rating = AudibleRatingService.instance.peek(widget.itemId);
      _load();
    }
  }

  Future<void> _load() async {
    final svc = AudibleRatingService.instance;
    // 1) show whatever is cached, instantly.
    final cached = await svc.loadCached(widget.itemId);
    if (!mounted) return;
    if (cached != null) setState(() => _rating = cached);
    // 2) refresh if stale/missing, then update in place.
    final fresh = await svc.resolve(
      itemId: widget.itemId,
      asin: widget.asin,
      title: widget.title,
      author: widget.author,
      narrator: widget.narrator,
      durationMs: widget.durationMs,
      region: widget.region,
    );
    if (!mounted || fresh == null) return;
    setState(() => _rating = fresh);
  }

  @override
  Widget build(BuildContext context) {
    final r = _rating;
    if (r == null || !r.found || r.rating <= 0) {
      return const SizedBox.shrink();
    }
    final cs = Theme.of(context).colorScheme;
    final starColor = widget.color ?? const Color(0xFFF6A609); // Audible amber
    final textColor = widget.color ?? cs.onSurfaceVariant;

    return Row(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: widget.alignment,
      children: [
        ..._buildStars(r.rating, widget.starSize, starColor),
        const SizedBox(width: 6),
        Text(
          r.rating.toStringAsFixed(1),
          style: TextStyle(
            fontSize: widget.starSize * 0.85,
            fontWeight: FontWeight.w700,
            color: textColor,
          ),
        ),
        if (widget.showCount && r.count != null && r.count! > 0) ...[
          const SizedBox(width: 4),
          Text(
            '(${_formatCount(r.count!)})',
            style: TextStyle(
              fontSize: widget.starSize * 0.78,
              color: textColor.withOpacity(0.85),
            ),
          ),
        ],
      ],
    );
  }

  List<Widget> _buildStars(double rating, double size, Color color) {
    final full = rating.floor();
    final frac = rating - full;
    return List.generate(5, (i) {
      IconData icon;
      if (i < full) {
        icon = Icons.star_rounded;
      } else if (i == full && frac >= 0.25 && frac < 0.85) {
        icon = Icons.star_half_rounded;
      } else if (i == full && frac >= 0.85) {
        icon = Icons.star_rounded;
      } else {
        icon = Icons.star_outline_rounded;
      }
      return Icon(icon, size: size, color: color);
    });
  }

  String _formatCount(int n) {
    if (n >= 1000000) return '${(n / 1000000).toStringAsFixed(1)}M';
    if (n >= 1000) return '${(n / 1000).toStringAsFixed(n >= 10000 ? 0 : 1)}k';
    return '$n';
  }
}

File: app/src/main/java/com/bennybar/kitzi/ui/common/AudibleStars.kt
package com.bennybar.kitzi.ui.common

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.filled.StarBorder
import androidx.compose.material.icons.filled.StarHalf
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.bennybar.kitzi.data.audible.AudibleRatingService
import java.util.Locale
import kotlin.math.floor

/**
 * A cached Audible community star rating for a book: shows the cached value
 * immediately and refreshes in place when it's stale (>24h). Renders nothing
 * until a confident rating exists — matching the Flutter widget.
 */
@Composable
fun AudibleStars(
    itemId: String,
    title: String,
    author: String? = null,
    narrator: String? = null,
    durationMs: Long? = null,
    starSize: Dp = 16.dp,
    showCount: Boolean = true,
    modifier: Modifier = Modifier,
) {
    var rating by remember(itemId) { mutableStateOf(AudibleRatingService.peek(itemId)) }
    LaunchedEffect(itemId) {
        AudibleRatingService.loadCached(itemId)?.let { rating = it }
        AudibleRatingService.resolve(itemId, title, author, narrator, durationMs)?.let { rating = it }
    }

    val r = rating ?: return
    if (!r.found || r.rating <= 0.0) return

    val amber = Color(0xFFF6A609)
    val muted = MaterialTheme.colorScheme.onSurfaceVariant
    Row(modifier, horizontalArrangement = Arrangement.spacedBy(1.dp), verticalAlignment = Alignment.CenterVertically) {
        val full = floor(r.rating).toInt()
        val frac = r.rating - full
        val fullCount = if (frac >= 0.75) full + 1 else full
        val hasHalf = frac >= 0.25 && frac < 0.75
        for (i in 0 until 5) {
            when {
                i < fullCount -> Icon(Icons.Default.Star, null, tint = amber, modifier = Modifier.size(starSize))
                i == fullCount && hasHalf -> Icon(Icons.Default.StarHalf, null, tint = amber, modifier = Modifier.size(starSize))
                else -> Icon(Icons.Default.StarBorder, null, tint = amber.copy(alpha = 0.30f), modifier = Modifier.size(starSize))
            }
        }
        Text(
            String.format(Locale.US, "%.1f", r.rating),
            fontSize = (starSize.value * 0.85f).sp,
            fontWeight = FontWeight.Bold,
            color = muted,
            modifier = Modifier.padding(start = 6.dp),
        )
        if (showCount && (r.count ?: 0) > 0) {
            Text(
                "(${formatCount(r.count!!)})",
                fontSize = (starSize.value * 0.78f).sp,
                color = muted.copy(alpha = 0.85f),
                modifier = Modifier.padding(start = 4.dp),
            )
        }
    }
}

private fun formatCount(n: Int): String =
    if (n >= 1000) String.format(Locale.US, "%.1fk", n / 1000.0).replace(".0k", "k") else n.toString()
